### ISSUE TITLE
multiple `stopSequence`

### ISSUE THREAD

**Issue by djmango** [eastriverlee/LLM.swift#14]
**Describe the bug**
Some models, like mistral v0.2 instruct (mistral-7b-instruct-v0.2.Q4_K_M.gguf), do not always correctly spit out stop tokens. This results in infinite generation, the model rambling forever. The stop feature you implemented allows for manual interruption here, but there should be a more robust solution. Ollama accounts for this by allowing each model to define multiple stop sequences, and watches for them all.

**To Reproduce**
1. Trigger a generation with mistral-7b-instruct-v0.2.Q4_K_M.gguf (about 10% chance that this will occur, from my testing)
2. Model will generate forever, tying up system resources and providing output with progressive degradation in quality

**Expected behavior**
The model should stop at a logical point in its generation all the time.

**Screenshots**
<img width="694" alt="image" src="https://github.com/eastriverlee/LLM.swift/assets/28496988/084487c6-5966-42c0-ba83-c3aa9aabcfd0">
<img width="1257" alt="image" src="https://github.com/eastriverlee/LLM.swift/assets/28496988/266dd6d3-7b2f-4296-94bb-64f384f8564d">

**Desktop:**
- Chip: M1 Pro
- Memory: 34GB
- OS: macOS 14.2

**Additional context**
This is within my grav.ai application. I will work on a solution today and PR.

**Comment by eastriverlee** [comment]
this is not a bug, as it's just how LLM works for now and has nothing to do with the library. rather, it's a feature request.

can you give me a minimal code that can reproduce what you were experiencing?
find and set `seed` parameter to a certain number, and a input string combination that can reproduce it.

**Comment by djmango** [comment]
Trying to recreate but it is unreliable. Am now logging seed and params so will catch it next time I see it in a debug environment. In doing so though I've found a new interesting bug stemming from the context limit:

<img width="703" alt="image" src="https://github.com/eastriverlee/LLM.swift/assets/28496988/657ffe04-94e9-43ad-bef3-b71ad095f76f">

When generating, the generation freezes for a second while I assume `history` and other input to the context window is truncated or deleted, and then the generation of the answer seems to restart, not considering what has already been generated. It is significantly better than crashing though, thank you for solving that.

**Comment by eastriverlee** [comment]
that's not a bug. if it hits `maxTokenCount`, first it attempts to remove the history and then resume, but if there is no history it uses `scoup` function that scoups out the middle part of the output(that is getting too long), and then try to "complete" that scouped out output. `getResponse` will call `finishResponse` if it cannot meet EOS within `maxTokenCount` and `finishResponse` uses `getResponse` internally so it can create a loop.

```swift
extension [String] {
    mutating func scoup(_ count: Int) {
        guard 0 < count else { return }
        let firstIndex = count
        let lastIndex = count * 2
        self.removeSubrange(firstIndex..<lastIndex)
    }
}
```

the only time it cannot generate output is when there is no history, but input itself is too long.

**Comment by djmango** [comment]
Got it, two updates:

1. I was able to get the infinite generation to replicate, I will explore and report findings/solutions and commit to my branch:
```
SEEDED WITH 3819086369
PARAMS: llama_context_params(seed: 3819086369, n_ctx: 4098, n_batch: 4098, n_threads: 10, n_threads_batch: 10, rope_scaling_type: -1, rope_freq_base: 0.0, rope_freq_scale: 0.0, yarn_ext_factor: -1.0, yarn_attn_factor: 1.0, yarn_beta_fast: 32.0, yarn_beta_slow: 1.0, yarn_orig_ctx: 0, cb_eval: nil, cb_eval_user_data: nil, type_k: __C.ggml_type(rawValue: 1), type_v: __C.ggml_type(rawValue: 1), mul_mat_q: true, logits_all: false, embedding: false, offload_kqv: true)
MODEL: mistral-7b-instruct-v0.2.Q4_K_M.gguf
PROMPT: yo
```

2. I've refactored the codebase on my branch significantly. The purpose was to remove as many class globals as possible and implement a new context limit truncation method, I've copied it below. Those are breaking changes, and I understand if that's not your preferred software arch philosophy, so won't open a PR unless you confirm. I will do my best to merge your changes into my branch and keep things on-par as far as features go. If you'd like to merge/discuss further, happy to chat or call on whatever, my contact info is on my profile README.

```swift
    /// Returns list of tokens, encoded from the history, truncated to the maximum token count.
    private func truncateAndEncode(history: [Chat]) -> [Token] {
        var tokens = encode(preProcess(history))
        guard tokens.count > maxTokenCount else { return tokens }

        // Truncate content from the start of history and recount tokens
        var truncatedHistory = history

        var index = 0

        // Then remove the first chat until we have removed enough tokens
        while tokens.count > maxTokenCount, index < truncatedHistory.count {
            // let chatTokenCount = encode(truncatedHistory[index].content).count

            // Get the amount of content we need to remove
            let tokensToRemove = (tokens.count - maxTokenCount) + 50 // Add 50 to ensure we remove enough tokens
            let contentToRemove = min(
                tokensToRemove * 4,
                truncatedHistory[index].content.count
            ) // Multiply by 4 for average token

            if contentToRemove == truncatedHistory[index].content.count {
                truncatedHistory.remove(at: index)
                print("Removed all of chat \(index)")
                index -= 1
            } else {
                // Remove the content
                truncatedHistory[index].content.removeFirst(contentToRemove)
                print("Removed \(contentToRemove) characters from chat \(index)")
            }

            // Update tokens
            tokens = encode(preProcess(truncatedHistory))
            print("Tokens: \(tokens.count)")
            index += 1
        }

        return encode(preProcess(truncatedHistory))
    }
```

**Comment by eastriverlee** [comment]
<img width="627" alt="Screenshot 2024-02-01 at 18 26 52" src="https://github.com/eastriverlee/LLM.swift/assets/43613000/9588d969-a3af-4ff4-bccd-e38462279d4c">

```swift
func test() async {
    let bot = LLM(from: Bundle.main.url(forResource: "mistral-7b-instruct-v0.2.Q4_K_M", withExtension: "gguf")!, template: .mistral, seed: 3819086369)
    print(await bot.getCompletion(from: bot.template!.preProcess("yo", [])))
}
```
i did a quick test using your seed,
and it didn't cause any problem.

**Comment by eastriverlee** [comment]
i have put some thoughts on this for some time, and i concluded that this is not necessary.
- firstly, i suspect @djmango  wasn't using this library when he was experiencing this infinite generation, as the example he presented is not reproducible.
- secondly, if you find yourself in a position where you need multiple `stopSequence`s, that's probably a sign that's telling you to find a better model. ***`stopSequence` is there to safeguard the edge case when LLM fails to produce EOS token.***
- thirdly, if you really want to have multiple `stopSequence`s, you can do so by overriding default behavior of `open func respond(to input: String) async` without rewriting the library.
- lastly, you can just change `stopSequence` if you don't think the default one i provide for the templates are not working well, as i was lazily defining `stopSeqeucne` for popular models, because they don't need one mostly anyway. for example you can change the `stopSequence` of `Template.mistral` with `"[INST]"` instead of `"</s>"` and see what works better.
